Source/Core/Logger/Logger.swift

```diff
@@ -0,0 +1,302 @@
+        }
+    }
+
+    /// Filters log messages unless they are tagged with any of `tags`
+    public func filterUnless(tags: [String]) {
+        self.queue.async { [weak self] in
+            if let union = self?.allowedTags.union(tags) {
+                self?.allowedTags = union
+            }
+        }
+    }
+
+    /// Filters log messages if they are tagged with `tag`
+    public func filterIf(tag: String) {
+        _ = self.queue.async { [weak self] in
+            self?.ignoredTags.insert(tag)
+        }
+    }
+
+    /// Filters log messages if they are tagged with any of `tags`
+    public func filterIf(tags: [String]) {
+        self.queue.async { [weak self] in
+            if let union = self?.ignoredTags.union(tags) {
+                self?.ignoredTags = union
+            }
+        }
+    }
+
+    func log<T>(
+        level: LogLevel = .info,
+        _ object: @autoclosure () -> T,
+        tag: String,
+        force: Bool = false,
+        _ file: String = #file,
+        _ function: String = #function,
+        _ line: Int = #line
+    ) {
+        self.log(level: level, object, tags: [tag], force: force, file, function, line)
+    }
+
+    func log<T, S>(
+        level: LogLevel = .info,
+        from _: S?,
+        _ object: @autoclosure () -> T,
+        tags: [String] = [],
+        force: Bool = false,
+        _ file: String = #file,
+        _ function: String = #function,
+        _ line: Int = #line
+    ) {
+        self.log(
+            level: level,
+            object: object,
+            tags: tags,
+            force: force,
+            context: String(describing: S.self),
+            file: file,
+            function: function,
+            line: line
+        )
+    }
+
+    func log<T>(
+        level: LogLevel = .info,
+        _ object: @autoclosure () -> T,
+        tags: [String] = [],
+        force: Bool = false,
+        _ file: String = #file,
+        _ function: String = #function,
+        _ line: Int = #line
+    ) {
+        self.log(
+            level: level,
+            object: object,
+            tags: tags,
+            force: force,
+            context: nil,
+            file: file,
+            function: function,
+            line: line
+        )
+    }
+
+    private func log<T>(
+        level: LogLevel = .info,
+        object: @autoclosure () -> T,
+        tags explicitTags: [String] = [],
+        force: Bool,
+        context: String?,
+        file: String,
+        function: String,
+        line: Int
+    ) {
+        #if !DEBUG
+            guard level != .debug else {
+                return
+            }
+        #endif
+
+        let thread = Thread.isMainThread ? "UI" : "BG"
+        let threadID = pthread_mach_thread_np(pthread_self())
+        let timestamp = self.startTime.elapsed
+        let string = "\(object())"
+
+        self.dispatchGroup.enter()
+        self.queue.async { [weak self] in
+            guard let strongSelf = self else { return }
```

Comment:
What happens to dispatchGroup when you call `enter()` and self is destroyed? Does it care, like holding on to something so it won't get destroyed?

Reply:
I would hope the destructor of DispatchGroup released it's resources. leave afaik just decrements a counter but I'm not sure of the implementation details. Couldn't find docs warning against anything weird here either. You know something I don't? 😛

Reply:
Only think I know is this:
```
 Calling this function indicates another block has joined the group through
 a means other than dispatch_group_async(). Calls to this function must be
 balanced with dispatch_group_leave().
```

Don't know what they mean with _must_ in:
 _Calls to this function must be balanced with dispatch_group_leave()_

Reply:
From this link https://github.com/apple/swift-corelibs-libdispatch/blob/master/dispatch/group.h#L237